src/tqec/circuit/detectors/match_utils/cover.py

```diff
@@ -0,0 +1,265 @@
+import typing as ty
+
+import pysat.solvers
+from tqec.circuit.detectors.boundary import BoundaryStabilizer
+from tqec.circuit.detectors.match_utils.sat import (
+    encode_pauli_string_commuting_cover_sat_problem_in_solver,
+    encode_pauli_string_exact_cover_sat_problem_in_solver,
+)
+from tqec.circuit.detectors.pauli import PauliString
+
+
+def _all_pauli_string_combination_results(
+    pauli_strings: list[PauliString],
+) -> ty.Iterator[tuple[list[bool], PauliString]]:
+    """Iterate over all the possible Pauli string products.
+
+    This function iterates over all the `2^{len(pauli_strings)}` products
+    that may be generated by either picking or not picking any of the
+    provided Pauli strings.
+
+    It is efficient in the sense that it performs the minimum number of
+    Pauli products required to do so.
+
+    Args:
+        pauli_strings: a list of Pauli strings that will be considered
+            in the returned products.
+
+    Yields:
+        all the 2^{len(pauli_strings)} possible Pauli products, one by one.
+    """
+    yield from _all_pauli_string_combination_results_impl(
+        pauli_strings, PauliString({})
+    )
+
+
+def _all_pauli_string_combination_results_impl(
```

Comment:
It is functionally equivalent. The main difference is that this implementation is re-using common Pauli string products, avoiding to perform the same Pauli string product more than once (which would be the case with the functionally equivalent method you provided, as it would need to compute `P0P1P2` twice to compute `P0P1P2P4P5` and `P0P1P2P4`).